Tutorial tracking UI

lobrien opened the pull request:
## Related Issues/PRs

[Jira task DOC-9761 ](https://databricks.atlassian.net/browse/DOC-9761)

## What changes are proposed in this pull request?

New tutorial doc: Compare and choose with the MLflow Tracking UI

## How is this patch tested?

Local build and review 

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [X ] Manual tests (describe details, including test results, below)

<!--
Please describe how you confirmed the proposed feature/bug-fix/change works here. For example, if you fixed an MLflow client API, you could attach the code that didn't work prior to the fix but works now, or if you added a new feature on MLflow UI, you could attach a video that demonstrates the feature.
-->

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ X] Yes. Give a description of this change to be included in the release notes for MLflow users.

New tutorial on the MLflow Tracking UI. Companion / follow-up to tracking tutorial (https://github.com/mlflow/mlflow/pull/8901). 

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ x] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

Commit:
Initial commit: Tutorial Tracking UI

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Intro

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
First draft

Contains back-references to tracking.rst

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

mlflow-automation commented:
<!-- documentation preview -->


Documentation preview for 3eb59dd13dea38bdcbfd6068cda91627b049a9e2 will be available [here](https://output.circle-artifacts.com/output/job/47adb49e-5887-447e-b958-416d0185cc6d/artifacts/0/docs/build/html/index.html) when [this CircleCI job](https://circleci.com/gh/mlflow/mlflow/55678) completes successfully.

<details>
<summary>More info</summary>

- Ignore this comment if this PR does not change the documentation.
- It takes a few minutes for the preview to be available.
- The preview is updated when a new commit is pushed to this PR.
- This comment was created by https://github.com/mlflow/mlflow/actions/runs/5480248732.

</details>

lobrien commented:
I think the build break is that this references **tutorial-tracking.rst** which is not yet merged. Not sure if I should pull the links temporarily or just leave as-is. Gonna' leave as-is for the moment.

BenWilson2 commented on docs/source/tutorial-tracking-ui.rst:
```diff
@@ -0,0 +1,140 @@
+If you have an instance of the MLflow Tracking Server running, you can access the Tracking UI by browsing to its location (for instance, ``http://localhost:5000``).
+
+You can also run the Tracking UI from the command-line with the ``mlflow ui`` command. It accepts many of the same arguments as the Tracking Server (``--port``, ``--host``, ``--backend-store-uri``, etc.). By default, ``mlflow ui`` will look for the ``MLFLOW_TRACKING_URI`` environment variable and use that as the backend store URI. If you do not set this environment variable and do not use the ``- ...
+
```
Can we add an example here of setting some directory on the local FS as the `backend-store-uri` that houses the run data that we've recorded and the artifact data stored in `artifacts-destination`, then show the actual cli command to start the UI pointing to that location, also showing a localhost with the correct port opened for the UI server (just to make it a copyable instruction for users)?

lobrien replied:
So two instances, the server on :5000 (say) and the ui on :5001 (say) but instead of both using `MLFLOW_TRACKING_URI`, both doing it with `--backend-store-uri` and `--artifacts-destination` args?

BenWilson2 replied:
that would be great :)

Commit:
Cherry-pick indexes; dummy file for tracking

TODO: Merge tracking tutorial (tutorial-tracking.rst) in other PR and clobber this placeholder.

Doc exists just so sphinx cross-links, including ToC, work.

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Add `mlflow gc` code snippet

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Reverted-out gateway/index.rst

Getting build error on this file, which was cherry-picked in

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

BenWilson2 commented on docs/source/_static/images/ui-tutorial/tracking_ui.png:
The indicators are great

BenWilson2 commented on docs/source/tutorial-tracking-ui.rst:
```diff
@@ -0,0 +1,144 @@
+
+.. important:: 
+
+   You should complete the :ref:`quickstart`, or even better, the :ref:`tutorial-tracking` before starting this how-to. This document assumes that you have several runs for which you've used the MLflow Tracking API to log parameters, metrics, and artifacts.
```
```suggestion
   You should complete the :ref:`quickstart`, or even better, the :ref:`tutorial-tracking` before starting this how-to. This document assumes that you have several runs for which you've used the MLflow Tracking API to log tags, parameters, metrics, and artifacts.
```

BenWilson2 commented on docs/source/tutorial-tracking-ui.rst:
```diff
@@ -0,0 +1,144 @@
+
+If you do not have an instance of the MLflow Tracking Server already running:
+
+1. Navigate to the directory containing the ``mlruns/`` subdirectory you want to visualize. For instance, the Quickstarts and :ref:`tutorial-tracking` expect you to be in the **examples/hyperparam** subdirectory of the cloned MLflow repository.
```
Can we link the `Quickstarts` to that section of the docs?

BenWilson2 commented on docs/source/tutorial-tracking-ui.rst:
```diff
@@ -0,0 +1,144 @@
+* Parameters | lr
+* Parameters | momentum
+
+Close the dropdown. You may have to horizontally scroll the runs table, but you can now see the chosen parameters and resulting metrics for each run. Usually, the best RMSE is slightly above ``0.49``. Note that the first run in the sweep has an ``lr`` of ``0``, so the model does not train beyond its random initialization; this usually results in an RMSE around ``0.89``.
```
```suggestion
Close the dropdown. You may have to horizontally scroll the runs table, but you can now see the chosen parameters and resulting metrics for each run. Usually, the best RMSE is slightly above ``0.49``. Note that the first run in the sweep has an ``lr`` of ``0``, so the model does not train beyond its random initialization; this usually results in an RMSE of around ``0.89``.
```

BenWilson2 commented on docs/source/tutorial-tracking-ui.rst:
```diff
@@ -0,0 +1,144 @@
+Compare runs visually with the Chart view
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+While you may have a good idea of the best run from the **Table view**, it is often useful to compare runs visually. Select the **Chart view** tab to switch to the chart view.
```
or when there are a LOT of runs (running hyperparameter tuning in a staging environment when tuning, I've generated 10's of thousands of runs in a single experiment :) ) definitely not something that I want to read in table format!

BenWilson2 commented on docs/source/tutorial-tracking-ui.rst:
```diff
@@ -0,0 +1,144 @@
+
+On this page, the main body is split between the table showing the runs in the experiment and a column which you can populate with various charts comparing runs. 
+
+The **Parallel Coordinates** chart is particularly useful for insight into the results of varying parameters. If there is no **Parallel Coordinates** chart in the right-hand column, click the **Add chart** button to add one. Select **Configure chart**. In the **Params** list, select **lr** and **momentum**. In the **Metrics** list, select **val_rmse**. Click **Save changes**.
```
```suggestion
The **Parallel Coordinates** chart is particularly useful for insight into the results of varying parameters, such as when performing hyperparameter tuning. If there is no **Parallel Coordinates** chart in the right-hand column, click the **Add chart** button to add one. Select **Configure chart**. In the **Params** list, select **lr** and **momentum**. In the **Metrics** list, select **val_rmse**. Click **Save changes**.
```

BenWilson2 commented on docs/source/tutorial-tracking-ui.rst:
```diff
@@ -0,0 +1,144 @@
+.. code-block:: bash
+
+  mlflow gc
+
```
Can we add a note here that mentions that if the underlying artifact has been manually deleted (someone went into the file store and deleted the directory or files), a warning will be issued for the invalid location while gc continues to operate on the remaining elements that have been marked for deletion in the tracking server?

BenWilson2 commented on docs/source/tutorial-tracking-ui.rst:
```diff
@@ -0,0 +1,144 @@
+
+  mlflow gc
+
+You may also delete a run using the CLI command ``mlflow run delete --run-id <run_id>`` or the Python API ``mlflow.delete_run(run_id : str)->None``.
```
(add note that `mlflow gc` will still need to be run if using the sdk or the cli commands in order to delete the saved artifact files)

BenWilson2 commented on docs/source/tutorial-tracking-ui.rst:
```diff
@@ -0,0 +1,144 @@
+.. image:: _static/images/ui-tutorial/run-details.png
+   :width: 100%
+
+In particular, note the **Run ID** (1), which is the primary key for the run. The run's name(for instance, ``stylish-ox-217``) is not guaranteed to be unique and cannot be used to identify a run in a program or at the command-line.
```
Well, the chances of collision are pretty low ;) 
A good callout, though! Can we mention that setting the run name during run creation (whether through the explicit api call `mlflow.start_run()` or through the run context `with mlflow.start_run()` that the run name can be set, overriding the generated fun name that we create for users?

BenWilson2 commented on docs/source/tutorial-tracking-ui.rst:
```diff
@@ -0,0 +1,144 @@
+
+In particular, note the **Run ID** (1), which is the primary key for the run. The run's name(for instance, ``stylish-ox-217``) is not guaranteed to be unique and cannot be used to identify a run in a program or at the command-line.
+
+The details page contains a large amount of information about the run, including the run's start time, duration, and status, the source code entry point, the Git commit hash of the code at the time of the run, and the run's parameters, metrics, and tags. The **Artifacts** section shows the artifacts logged by the run.
```
Also worthwhile to point out that the `Artifacts` section also shows the input example (if generated), the inference signature, model inference dependencies (requirements), and the model Metadata associated with the run; all of which are either used for model loading, serving, or debugging?

BenWilson2 commented:
Great stuff! I followed the tutorial line by line without looking at the screenshots and even without the visual clues it was very clear what everything did and how to follow along in the UI to navigate. Well done!

harupy commented on docs/image-sources/ui-tutorial/mlflow_registry_transitions.psd:
Do we need these files? They are not small. I think they will slow down git clone once we push them.
